### ISSUE TITLE
Analyzer should not accept a typedef that refers to itself via bound

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#43790]
The following program is accepted by the analyzer but rejected by the CFE:
```dart
typedef F<T extends F<Null>> = T Function();

main() {}
```

The CFE's error message is: `Error: The typedef 'F' has a reference to itself.`

I believe the analyzer should reject this program, because of the following spec text (from the "Type Aliases" section):

> Any self reference in a type alias, either directly or recursively via another type declaration, is a compile-time error.

We will probably want to fix this issue either before or during our work on analyzer support for non-function type aliases.

**Comment by stereotype441** [comment]
After further experimentation it seems that the analyzer and the CFE have a number of differences in how they check for circularities.  I've filed a language issue, and I'll postpone further work on this until that language issue is resolved: https://github.com/dart-lang/language/issues/1261

**Comment by leafpetersen** [comment]
@eernstg do we have a resolution for this?

**Comment by eernstg** [comment]
I've proposed an update to the language specification in https://dart-review.googlesource.com/184460 which specifies the strict approach (where both body dependencies and bound dependencies give rise to a cyclicity error). This allows us to relax the rule later on (to allow bound dependencies), which would be a non-breaking change.

I've added a language test in https://dart-review.googlesource.com/c/sdk/+/186290, covering areas that weren't already covered for this topic.

The example [here](https://github.com/dart-lang/sdk/issues/43790#issue-721658378) is flagged as an error by both the analyzer and the CFE today.

The new test causes a crash in the CFE, cf. https://github.com/dart-lang/sdk/issues/45085, but the analyzer already passes the new test (and the existing ones), so there is no implementation effort associated with the language rule itself.
